disjoint-sets.lisp

```diff
@@ -28,27 +23,20 @@ Examples:
 (make-disjoint-sets 10)
 ;; => #(0 1 2 3 4 5 6 7 8 9)
 "
-  (let ((sets (make-array (list (or number-of-sets 0))
-                          :element-type 'integer
-                          :adjustable t
-                          :fill-pointer t)))
-    (when number-of-sets
-      (loop :for i :below number-of-sets
-            :do (setf (aref sets i) i)))
-    sets))
+  (let ((sets (make-array `(,arity) :element-type 'integer
+                          :adjustable t :fill-pointer t)))
+    (dotimes (i arity sets) (setf (aref sets i) i))))
 
 (defun disjoint-sets-add (sets)
-  "Add a new item into its own disjoint set. Return a new id.
+  "Add a new item into its own disjoint set. Return the new id.
 
 Example:
 
 (let ((id (disjoint-sets-add sets)))
   ;; SETS is modified
   ...)
 "
-  (let ((new-id  (length sets)))
```

Comment:
The previous version is more easily explainable (arguably).

I would prefer to keep it like it was even if the `let` is only there to give a name to an intermediate value.

Reply:
ok; it's probably also more idiomatic if you think about this code being a reduction of more complicated code that operates on objects that store lots of graph metadata, rather than only the equivalence indices.

Reply:
TL;DR: good point

---

another way of implementing this data structure is to use a hash-table with ids as keys and the set reprensentatives' id as value.

with that variant, ids can ben anything

I plan on implenting that in breeze, because I need it in the e-graph's implementation (IIUC what I have left to do there).